File: arduino-yun/arduino-IO-node/LCD.cpp
/******************************************************************************
  LCD.cpp - LCD library Version 1.11
  Copyright (C), 2015, Sparking Work Space. All right reserved.

 ******************************************************************************

  This library is free software; you can redistribute it and/or
  modify it under the terms of the GNU Lesser General Public
  License as published by the Free Software Foundation; either
  version 2.1 of the License, or (at your option) any later version.

  This library is build for LCD12864. Please do not use this library on
  any other devices, that could cause unpredictable damage to the unknow device.
  See the GNU Lesser General Public License for more details.

  You should have received a copy of the GNU Lesser General Public
  License along with this library; if not, write to the Free Software
  Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA  02110-1301  USA

*******************************************************************************/


#include "LCD.h"



/*const Font_Info_t fontInfo[LCD_NUM_OF_FONT] =
  {
    Font_6x8, 6, 8,
    Font_6x12, 6, 12,
    Font_8x16_1, 8, 16,
    Font_8x16_2, 8, 16,
    Font_10x20, 10, 20,
    Font_12x24, 12, 24,
    Font_16x32, 16, 32,
  };*/

/********************* The start of LCD basic driver APIs *******************/

LCD::LCD(void) {};

/**************************************************************
      I2C init.
***************************************************************/
void LCD::Init(void)
{
#ifndef NOLCD

#ifndef I2CLIB
  I2C_BUS_Init();
#else
  I2c.begin();
#endif
#endif
}

/**************************************************************
      Read one byte from device register.
***************************************************************/
uint8_t LCD::ReadByteFromReg(enum LCD_RegAddress regAddr)
{
#ifndef NOLCD

#if 0
  int8_t waitTime;
  I2C_BUS_BeginTransmission(LCD_ADDRESS);
  I2C_BUS_WriteByte(regAddr);
  I2C_BUS_EndTransmission();

  I2C_BUS_RequestFrom((int16_t)LCD_ADDRESS, 1);
  for (waitTime = 10; !I2C_BUS_Available() && waitTime; waitTime--)
    __asm__("nop\n\t");
  if (I2C_BUS_Available())
  {
    I2C_BUS_EndTransmission();
    return I2C_BUS_ReadByte();
  }
  else
    return 0;
#else
  uint8_t val = 0;
 /* if (I2c.read(LCD_ADDRESS,regAddr,1,&val) == 0)
    return val;
  else
    return 0;*/
   
/*    I2c.start();
    I2c.sendAddress(LCD_ADDRESS);
    I2c.sendByte(regAddr);
    I2c.stop();
    val = I2c.receiveByte(0);*/
    return val;
    
#endif
#else
    return 0;
#endif
}

/**************************************************************
      Write one byte to device register.
***************************************************************/
void LCD::WriteByteToReg(enum LCD_RegAddress regAddr, uint8_t buf)
{
#ifndef NOLCD
#ifndef I2CLIB
  I2C_BUS_BeginTransmission(LCD_ADDRESS);
  I2C_BUS_WriteByte(regAddr);
  I2C_BUS_WriteByte(buf);
  I2C_BUS_EndTransmission();
#else
  I2c.write(LCD_ADDRESS,regAddr,buf);
#endif
#endif
}



/**************************************************************
      Write multiple bytes to device register.
***************************************************************/
void LCD::WriteSeriesToReg(enum LCD_RegAddress regAddr, const uint8_t *buf, uint8_t length)
{
#ifndef NOLCD
#ifndef I2CLIB
  uint8_t i;
  I2C_BUS_BeginTransmission(LCD_ADDRESS);
  I2C_BUS_WriteByte(regAddr);
  for (i = 0; i < length; i++)
  {
    I2C_BUS_WriteByte(buf[i]);
  }
  I2C_BUS_EndTransmission();    // stop transmitting
#else
  I2c.write(LCD_ADDRESS,regAddr,buf,length);
#endif
#endif
}

/********************* The end of LCD basic driver APIs *******************/

#ifdef  SUPPORT_FULL_API_LIB
#if  SUPPORT_FULL_API_LIB == TRUE

void LCD::CharGotoXY(uint8_t x, uint8_t y)
{
#ifndef NOLCD
  uint8_t buf[2];
  buf[0] = x;
  buf[1] = y;
  WriteSeriesToReg(CharXPosRegAddr, buf, 2);
#endif
}

void LCD::ReadRAMGotoXY(uint8_t x, uint8_t y)
{
#ifndef NOLCD
  uint8_t buf[2];
  buf[0] = x;
  buf[1] = y;
  WriteSeriesToReg(ReadRAM_XPosRegAddr, buf, 2);
#endif
}



void LCD::WriteRAMGotoXY(uint8_t x, uint8_t y)
{
#ifndef NOLCD
  uint8_t buf[2];
  buf[0] = x;
  buf[1] = y;
  WriteSeriesToReg(WriteRAM_XPosRegAddr, buf, 2);
#endif
}



void LCD::FontModeConf(enum LCD_FontSort font, enum LCD_FontMode mode, enum LCD_CharMode cMode)
{
#ifndef NOLCD
  WriteByteToReg(FontModeRegAddr, cMode | mode | font);
#endif
}

void LCD::DispCharAt(char buf, uint8_t x, uint8_t y)
{
#ifndef NOLCD
  CharGotoXY(x, y);
  WriteByteToReg(DisRAMAddr, buf);
#endif
}

void LCD::DispStringAt(char *buf, uint8_t x, uint8_t y)
{
#ifndef NOLCD
  CharGotoXY(x, y);
  for (; *buf; buf++)
    WriteByteToReg(DisRAMAddr, *buf);
#endif
}

const uint8_t fontYsizeTab[LCD_NUM_OF_FONT] = {8, 12, 16, 16, 20, 24, 32};

inline size_t LCD::write(uint8_t value)
{
#ifndef NOLCD
  uint8_t Y_Present, Y_New, Fontsize_Y, fontIndex;
  switch (value)
  {
    case 0x0d: break;
    case 0x0a:
      Y_Present = ReadByteFromReg(CharYPosRegAddr);
      fontIndex = ReadByteFromReg(FontModeRegAddr) & 0x0f;

      if (Y_Present + 2 * fontYsizeTab[fontIndex] <= LCD_Y_SIZE_MAX)
      {
        Y_New = fontYsizeTab[fontIndex] + Y_Present;
        CharGotoXY(0, Y_New);
      }
      else
        CharGotoXY(0, 0);
      break;
    case 0x09:
      WriteSeriesToReg(DisRAMAddr, (uint8_t *)"  ", 2);
      break;
    default: WriteByteToReg(DisRAMAddr, value);
  }
#endif
  return 1; // assume sucess
}


void LCD::CursorConf(enum LCD_SwitchState swi, uint8_t freq)
{
#ifndef NOLCD
  WriteByteToReg(CursorConfigRegAddr, (char)(swi << 7) | freq);
#endif
}

void LCD::CursorGotoXY(uint8_t x, uint8_t y, uint8_t width, uint8_t height)
{
#ifndef NOLCD
  uint8_t buf[4];
  buf[0] = x;
  buf[1] = y;
  buf[2] = width;
  buf[3] = height;
  WriteSeriesToReg(CursorXPosRegAddr, buf, 4);
#endif
}

#ifdef  SUPPORT_2D_GRAPHIC_LIB
#if  SUPPORT_2D_GRAPHIC_LIB == TRUE

void LCD::DrawDotAt(uint8_t x, uint8_t y, enum LCD_ColorSort color)
{
#ifndef NOLCD
  uint8_t buf[2];
  if (x < 128 && y < 64)
  {
    buf[0] = x;
    buf[1] = (uint8_t)(color << 7) | y;
    WriteSeriesToReg(DrawDotXPosRegAddr, buf, 2);
  }
#endif
}

void LCD::DrawHLineAt(uint8_t startX, uint8_t endX, uint8_t y, enum LCD_ColorSort color)
{
#ifndef NOLCD
  DrawLineAt(startX, endX, y, y, color);
#endif
}

void LCD::DrawVLineAt(uint8_t startY, uint8_t endY, uint8_t x, enum LCD_ColorSort color)
{
#ifndef NOLCD
  DrawLineAt(x, x, startY, endY, color);
#endif
}

void LCD::DrawLineAt(uint8_t startX, uint8_t endX, uint8_t startY, uint8_t endY, enum LCD_ColorSort color)
{
#ifndef NOLCD
  uint8_t buf[4];
  if (endY < 64)
  {
    buf[0] = startX;
    buf[1] = endX;
    buf[2] = startY;
    buf[3] = (uint8_t)(color << 7) | endY;
    WriteSeriesToReg(DrawLineStartXRegAddr, buf, 4);
  }
#endif
}

void LCD::DrawRectangleAt(uint8_t x, uint8_t y, uint8_t width, uint8_t height, enum LCD_DrawMode mode)
{
#ifndef NOLCD
  uint8_t buf[5];
  buf[0] = x;
  buf[1] = y;
  buf[2] = width;
  buf[3] = height;
  buf[4] = mode;
  WriteSeriesToReg(DrawRectangleXPosRegAddr, buf, 5);
#endif
}

void LCD::DrawCircleAt(int8_t x, int8_t y, uint8_t r, enum LCD_DrawMode mode)
{
#ifndef NOLCD
  uint8_t buf[4];
  if (x < 128 && y < 64 && r < 64)
  {
    buf[0] = x;
    buf[1] = y;
    buf[2] = r;
    buf[3] = mode;
    WriteSeriesToReg(DrawCircleXPosRegAddr, buf, 4);
  }
#endif
}


#endif
#endif


void LCD::DisplayConf(enum LCD_DisplayMode mode)
{
#ifndef NOLCD
  DisplayConfigRegVal = mode;
  WriteByteToReg(DisplayConfigRegAddr, mode);
#endif
}

void LCD::WorkingModeConf(enum LCD_SwitchState logoSwi, enum LCD_SwitchState backLightSwi, enum LCD_WorkingMode mode)
{
#ifndef NOLCD
  WriteByteToReg(WorkingModeRegAddr, 0x50 | (uint8_t)(logoSwi << 3) | (uint8_t)(backLightSwi << 2) | mode);
#endif
}

void LCD::BacklightConf(enum LCD_SettingMode mode, uint8_t buf)
{
#ifndef NOLCD
  if (buf > 0x7f)
    buf = 0x7f;
  WriteByteToReg(BackLightConfigRegAddr, mode | buf);
#endif
}

void LCD::ContrastConf(enum LCD_SettingMode mode, uint8_t buf)
{
#ifndef NOLCD
  if (buf > 0x3f)
    buf = 0x3f;
  WriteByteToReg(ContrastConfigRegAddr, mode | buf);
#endif
}

void LCD::DeviceAddrEdit(uint8_t newAddr)
{
#ifndef NOLCD
  uint8_t buf[2];
  buf[0] = 0x80;
  buf[1] = newAddr;
  WriteSeriesToReg(DeviceAddressRegAddr, buf, 2);
#endif
}

void LCD::CleanAll(enum LCD_ColorSort color)
{
#ifndef NOLCD
  uint8_t buf;


  buf = DisplayConfigRegVal ;
  if (color == WHITE)
    WriteByteToReg(DisplayConfigRegAddr, buf | 0x60);
  else 
    WriteByteToReg(DisplayConfigRegAddr, (buf | 0x40) & 0xdf);

#endif //NO LCD
}


#endif
#endif

File: siemens-iot2000/peripherals/lcd.js
/*
  Copyright (C) Air Liquide S.A,  2017
  Author: Sébastien Lalaurette and Cyril Chapellier, La Factory, Creative Foundry
  This file is part of Predictable Farm project.

  The MIT License (MIT)

  Permission is hereby granted, free of charge, to any person obtaining a copy
  of this software and associated documentation files (the "Software"), to deal
  in the Software without restriction, including without limitation the rights
  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
  copies of the Software, and to permit persons to whom the Software is
  furnished to do so, subject to the following conditions:

  The above copyright notice and this permission notice shall be included in
  all copies or substantial portions of the Software.

  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
  THE SOFTWARE.
   
  See the LICENSE.txt file in this repository for more information.
*/

'use strict'

const config = require('../config/default')
const logger = require('../services/logger')
const scheduler = require('../services/scheduler-client')

const mraa = require('/usr/lib/node_modules/mraa')

logger.init('lcd')
logger.log('MRAA Version: %s ', mraa.getVersion())

var i2cDevice = new mraa.I2c(0)
i2cDevice.address(0x51)

const FontModeRegAddr = 1
const CharXPosRegAddr = 2
const CharYPosRegAddr = 3

const CursorConfigRegAddr = 16
const CursorXPosRegAddr = 17
const CursorYPosRegAddr = 18
const CursorWidthRegAddr = 19
const CursorHeightRegAddr = 20 // 8

const DisRAMAddr = 32 // 9
const ReadRAM_XPosRegAddr = 33
const ReadRAM_YPosRegAddr = 34
const WriteRAM_XPosRegAddr = 35
const WriteRAM_YPosRegAddr = 36

const DrawDotXPosRegAddr = 64 // 14
const DrawDotYPosRegAddr = 65

const DrawLineStartXRegAddr = 66
const DrawLineEndXRegAddr = 67
const DrawLineStartYRegAddr = 68
const DrawLineEndYRegAddr = 69

const DrawRectangleXPosRegAddr = 70 // 20
const DrawRectangleYPosRegAddr = 71
const DrawRectangleWidthRegAddr = 72
const DrawRectangleHeightRegAddr = 73
const DrawRectangleModeRegAddr = 74

const DrawCircleXPosRegAddr = 75
const DrawCircleYPosRegAddr = 76
const DrawCircleRRegAddr = 77
const DrawCircleModeRegAddr = 78

const DrawBitmapXPosRegAddr = 79
const DrawBitmapYPosRegAddr = 80
const DrawBitmapWidthRegAddr = 81
const DrawBitmapHeightRegAddr = 82 // 31

const DisplayConfigRegAddr = 128 // 32
const WorkingModeRegAddr = 129
const BackLightConfigRegAddr = 130
const ContrastConfigRegAddr = 131
const DeviceAddressRegAddr = 132

const ArrowLeft = 0x1c
const ArrowRight = 0x1d
const ArrowUp = 0x1e
const ArrowDown = 0x1f

const WHITE = 0x00
const BLACK = 0xff

const TRUE = 1
const FALSE = 0

const BLACK_BAC = 0x00
const BLACK_NO_BAC = 0x40
const WHITE_BAC = 0x80
const WHITE_NO_BAC = 0xc0

const DisplayMode_AllREV = 0x00
const DisplayMode_AllNOR = 0x80

const OFF = 0x00
const ON = 0x01

const Font_6x8 = 0x00
const Font_6x12 = 0x01
const Font_8x16_1 = 0x02
const Font_8x16_2 = 0x03
const Font_10x20 = 0x04
const Font_12x24 = 0x05
const Font_16x32 = 0x06

const FM_ANL_AAA = 0x00 // FM_AutoNewLine_AutoAddrAdd
const FM_ANL_MAA = 0x10 // FM_AanualNewLine_ManualAddrAdd
const FM_MNL_MAA = 0x30 // FM_ManualNewLine_ManualAddrAdd
const FM_MNL_AAA = 0x20 // FM_ManualNewLine_AutoAddrAdd

const WM_CharMode = 0x00 // Common mode, put char or dot
const WM_BitmapMode = 0x01
const WM_RamMode = 0x02 // 8Bit deal mode, deal with 8-bit RAM directly

const LOAD_TO_RAM = 0x00
const LOAD_TO_EEPROM = 0x80

var DisplayConfigRegVal = DisplayMode_AllREV // DisplayMode_AllNOR;

function ReadByteFromReg (regAddr) {
  return i2cDevice.readReg((regAddr))
}

function WriteByteToReg (regAddr, buf) {
  i2cDevice.writeReg(regAddr, buf)
}

function WriteSeriesToReg (regAddr, buf, length) {
  var data = new Buffer(length + 1)
  data[0] = regAddr
  var i
  for (i = 0; i < length; i++) {
    data[i + 1] = Number(buf[i])
  }
  i2cDevice.write(data)
}

function FontModeConf (font, mode, cMode) {
  WriteByteToReg(FontModeRegAddr, cMode | mode | font)
}

function CharGotoXY (x, y) {
  var buf = [x, y]
  WriteSeriesToReg(CharXPosRegAddr, buf, 2)
}

function DispCharAt (buf, x, y) {
  CharGotoXY(x, y)
  WriteByteToReg(DisRAMAddr, buf.charCodeAt(0))
}

function DispCursorAt (cur, x, y) {
  CharGotoXY(x, y)
  WriteByteToReg(DisRAMAddr, cur)
}

function DispStringAt (buf, x, y) {
  // logger.log(x + " " + y + " " + buf);
  CharGotoXY(x, y)
  for (var i = 0; i < buf.length; i++) {
    WriteByteToReg(DisRAMAddr, buf.charCodeAt(i))
  }
}

function print (string) {
  var Y_Present, Y_New, Fontsize_Y, fontIndex
  for (var i = 0; i < string.length; i++) {
    switch (string.charCodeAt(i)) {
      case 0x0d: break
      case 0x0a:
        Y_Present = ReadByteFromReg(CharYPosRegAddr)
        fontIndex = ReadByteFromReg(FontModeRegAddr) & 0x0f

        if (Y_Present + 2 * fontYsizeTab[fontIndex] <= LCD_Y_SIZE_MAX) {
          Y_New = fontYsizeTab[fontIndex] + Y_Present
          CharGotoXY(0, Y_New)
        } else { CharGotoXY(0, 0) }
        break
      case 0x09:
        WriteSeriesToReg(DisRAMAddr, '  ', 2)
        break
      default: WriteByteToReg(DisRAMAddr, string.charCodeAt(i))
    }
  }
  return 1 // assume sucess
}

function DisplayConf (mode) {
  DisplayConfigRegVal = mode
  WriteByteToReg(DisplayConfigRegAddr, mode)
}

function WorkingModeConf (logoSwi, backLightSwi, mode) {
  WriteByteToReg(WorkingModeRegAddr, 0x50 | (logoSwi << 3) | (backLightSwi << 2) | mode)
}

function BacklightConf (mode, buf) {
  if (buf > 0x7f) { buf = 0x7f }
  WriteByteToReg(BackLightConfigRegAddr, mode | buf)
}

function ContrastConf (mode, buf) {
  if (buf > 0x3f) { buf = 0x3f }
  WriteByteToReg(ContrastConfigRegAddr, mode | buf)
}

function CleanAll (color) {
  var buf = DisplayConfigRegVal
  if (color == WHITE) { WriteByteToReg(DisplayConfigRegAddr, buf | 0x60) } else { WriteByteToReg(DisplayConfigRegAddr, (buf | 0x40) & 0xdf) }
}

function DrawDotAt (x, y, color) {
  var buf = new Buffer(2)
  if (x < 128 && y < 64) {
    buf[0] = x
    buf[1] = (color << 7) | y
    WriteSeriesToReg(DrawDotXPosRegAddr, buf, 2)
  }
}

function DrawHLineAt (startX, endX, y, color) {
  DrawLineAt(startX, endX, y, y, color)
}

function DrawVLineAt (startY, endY, x, color) {
  DrawLineAt(x, x, startY, endY, color)
}

function DrawLineAt (startX, endX, startY, endY, color) {
  var buf = new Buffer(4)
  if (endY < 64) {
    buf[0] = startX
    buf[1] = endX
    buf[2] = startY
    buf[3] = (uint8_t)(color << 7) | endY
    WriteSeriesToReg(DrawLineStartXRegAddr, buf, 4)
  }
}

function DrawRectangleAt (x, y, width, height, mode) {
  var buf = new Buffer(5)
  buf[0] = x
  buf[1] = y
  buf[2] = width
  buf[3] = height
  buf[4] = mode
  WriteSeriesToReg(DrawRectangleXPosRegAddr, buf, 5)
}

function DrawCircleAt (x, y, r, mode) {
  var buf = new Buffer(4)
  if (x < 128 && y < 64 && r < 64) {
    buf[0] = x
    buf[1] = y
    buf[2] = r
    buf[3] = mode
    WriteSeriesToReg(DrawCircleXPosRegAddr, buf, 4)
  }
}

function initialScreen () {
  WorkingModeConf(OFF, ON, WM_CharMode)
  FontModeConf(Font_6x8, FM_ANL_AAA, BLACK_BAC)
  CleanAll(WHITE)
  DispCharAt('X', 60, 2)
  DispStringAt('Air liquide', 20, 30)
}

var x = 0
var y = 0
var busy = false
function lcd_print_now (cmd_process) {
  // logger.log('lcd_print_now ' + cmd_process.length)
  if (busy) {
    logger.log('lcd_print_now busy')
    return
  }
  if (cmd_process.length > 0) {
    busy = true
    /*  var copy = [];
        copy = cmd_process; */
    var cmd = cmd_process.shift()
    var text = null
    var posx = 0
    var posy = 0

    // Reset fontmode in case we are back from a power cycle of the i2c bus
    FontModeConf(Font_6x8, FM_ANL_AAA, BLACK_BAC)

    if (cmd.hasOwnProperty('clean')) {
      CleanAll(WHITE)
    }
    if (cmd.hasOwnProperty('x')) {
      posx = Number(cmd.x)
    }
    if (cmd.hasOwnProperty('y')) {
      posy = Number(cmd.y)
    }
    if (cmd.hasOwnProperty('print')) {
      text = cmd.print
      if (cmd.hasOwnProperty('print_x')) {
        posx = Number(cmd.print_x)
      }
      if (cmd.hasOwnProperty('print_y')) {
        posy = Number(cmd.print_y)
      }
      if (text) DispStringAt(text, posx, posy)
      if (cmd.hasOwnProperty('print_box')) {
        if (text) DrawRectangleAt(posx - Number(cmd.print_box), posy - Number(cmd.print_box), (text.length * 6) + Number(cmd.print_box) * 2, 8 + Number(cmd.print_box) * 2, 0)
      }
      if (cmd.hasOwnProperty('print_cursor')) {
        if (text) DispCursorAt((cmd.print_cursor), posx - 8, posy)
        else DispCursorAt((cmd.print_cursor), posx, posy)
      }
    }
    if (cmd.hasOwnProperty('circle_radius')) {
      DrawCircleAt(posx, posy, Number(cmd.circle_radius), 0)
    }
    busy = false
  }
}

var local_service = scheduler.init({
  mode: scheduler.constants.MODE_QUEUE,
  module_name: 'lcd',
  onInit: () => {},
  onData: () => {},
  onConnect: () => {},
  onDisconnect: initialScreen,
  onSchedule: lcd_print_now
})

initialScreen()

/*
// DEBUG ONLY
var later = setInterval(() => {
  lcd_print_now()
}, 20)
*/
